File: packages/imsg-bridge/src/crypto.ts
/**
 * End-to-end encryption for the iMessage bridge tunnel.
 *
 * Protocol:
 *   1. Both sides generate ephemeral X25519 key pairs
 *   2. Exchange public keys during auth handshake
 *   3. Derive shared secret via ECDH
 *   4. Derive AES-256-GCM key via HKDF-SHA256
 *   5. Every frame encrypted with unique random IV
 *
 * Wire format (binary): [12-byte IV][ciphertext][16-byte auth tag]
 */

import {
  generateKeyPairSync,
  diffieHellman,
  hkdfSync,
  randomBytes,
  createCipheriv,
  createDecipheriv,
  createPublicKey,
  type KeyObject,
} from 'node:crypto';

// ── Key Exchange ────────────────────────────────────────────────────────────

export interface KeyPair {
  publicKey: KeyObject;
  privateKey: KeyObject;
}

/** Generate an ephemeral X25519 key pair for ECDH. */
export function generateKeyPair(): KeyPair {
  const { publicKey, privateKey } = generateKeyPairSync('x25519');
  return { publicKey, privateKey };
}

/** Export public key to raw 32-byte Buffer for wire transfer. */
export function exportPublicKey(key: KeyObject): Buffer {
  // DER format for X25519 public key: 12-byte header + 32-byte key
  const der = key.export({ type: 'spki', format: 'der' });
  return Buffer.from(der.subarray(12));
}

/** Import raw 32-byte public key from wire. */
export function importPublicKey(raw: Buffer): KeyObject {
  // Wrap raw 32 bytes in X25519 SPKI DER header
  const header = Buffer.from('302a300506032b656e032100', 'hex');
  const der = Buffer.concat([header, raw]);
  return createPublicKey({ key: der, format: 'der', type: 'spki' });
}

/** Derive a shared AES-256 key from local private + remote public via ECDH + HKDF. */
export function deriveSessionKey(localPrivate: KeyObject, remotePublic: KeyObject): Buffer {
  const sharedSecret = diffieHellman({
    privateKey: localPrivate,
    publicKey: remotePublic,
  });

  const salt = Buffer.from('botmem-imsg-tunnel-v1', 'utf-8');
  const info = Buffer.from('aes-256-gcm-session-key', 'utf-8');

  const derived = hkdfSync('sha256', sharedSecret, salt, info, 32);
  return Buffer.from(derived);
}

// ── Symmetric Encryption ────────────────────────────────────────────────────

const IV_LENGTH = 12;
const TAG_LENGTH = 16;

/** Encrypt plaintext with AES-256-GCM. Returns [IV (12) | ciphertext | tag (16)]. */
export function encrypt(key: Buffer, plaintext: string): Buffer {
  const iv = randomBytes(IV_LENGTH);
  const cipher = createCipheriv('aes-256-gcm', key, iv);

  const encrypted = Buffer.concat([cipher.update(plaintext, 'utf-8'), cipher.final()]);
  const tag = cipher.getAuthTag();

  return Buffer.concat([iv, encrypted, tag]);
}

/** Decrypt AES-256-GCM payload. Input: [IV (12) | ciphertext | tag (16)]. */
export function decrypt(key: Buffer, payload: Buffer): string {
  if (payload.length < IV_LENGTH + TAG_LENGTH) {
    throw new Error('Encrypted payload too short');
  }

  const iv = payload.subarray(0, IV_LENGTH);
  const tag = payload.subarray(payload.length - TAG_LENGTH);
  const ciphertext = payload.subarray(IV_LENGTH, payload.length - TAG_LENGTH);

  const decipher = createDecipheriv('aes-256-gcm', key, iv);
  decipher.setAuthTag(tag);

  const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]);

  return decrypted.toString('utf-8');
}

// ── Convenience ─────────────────────────────────────────────────────────────

/** Encrypt a JSON-serializable object. */
export function encryptJson(key: Buffer, data: unknown): Buffer {
  return encrypt(key, JSON.stringify(data));
}

/** Decrypt a payload and parse as JSON. */
export function decryptJson<T = unknown>(key: Buffer, payload: Buffer): T {
  return JSON.parse(decrypt(key, payload)) as T;
}

File: packages/apple-bridge-engine/src/tunnel/crypto.rs
//! Tunnel crypto — MUST match `packages/apple-bridge/src/crypto.ts` and the
//! server (`apple-tunnel.service.ts`) exactly. See PROTOCOL.md §4.
//!
//!   X25519 (raw 32-byte keys on the wire)
//!   → ECDH shared secret
//!   → HKDF-SHA256(salt="botmem-apple-tunnel-v1", info="aes-256-gcm-session-key", 32)
//!   → AES-256-GCM, frame = [IV(12)] [ciphertext] [tag(16)]
//!
//! NB: the Node side wraps/unwraps a 12-byte X25519 SPKI DER header around the
//! raw key purely because its KeyObject API needs DER. The ECDH itself operates
//! on the raw 32 bytes, which is exactly what we put on the wire — so here we
//! work with raw bytes directly and never touch the DER header.

use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Nonce};
use hkdf::Hkdf;
use sha2::Sha256;
use x25519_dalek::{PublicKey, StaticSecret};

use super::{HKDF_INFO, HKDF_SALT, IV_LEN, TAG_LEN};

/// An ephemeral X25519 keypair for one tunnel connection.
pub struct KeyPair {
    secret: StaticSecret,
    pub public_raw: [u8; 32],
}

impl KeyPair {
    /// Generate a fresh ephemeral keypair.
    pub fn generate() -> Result<Self, CryptoError> {
        let mut seed = [0u8; 32];
        getrandom::getrandom(&mut seed).map_err(|_| CryptoError::Rng)?;
        // StaticSecret applies X25519 clamping on construction.
        let secret = StaticSecret::from(seed);
        let public = PublicKey::from(&secret);
        Ok(Self { secret, public_raw: *public.as_bytes() })
    }

    /// Derive the AES-256 session key from the remote raw 32-byte public key.
    pub fn derive_session_key(&self, remote_public_raw: &[u8]) -> Result<[u8; 32], CryptoError> {
        if remote_public_raw.len() != 32 {
            return Err(CryptoError::BadPublicKey);
        }
        let mut pk = [0u8; 32];
        pk.copy_from_slice(remote_public_raw);
        let remote = PublicKey::from(pk);
        let shared = self.secret.diffie_hellman(&remote);

        let hk = Hkdf::<Sha256>::new(Some(HKDF_SALT), shared.as_bytes());
        let mut okm = [0u8; 32];
        hk.expand(HKDF_INFO, &mut okm).map_err(|_| CryptoError::Hkdf)?;
        Ok(okm)
    }
}

/// Encrypt plaintext with AES-256-GCM. Output: `[IV(12)] [ciphertext] [tag(16)]`.
pub fn encrypt(key: &[u8; 32], plaintext: &[u8]) -> Result<Vec<u8>, CryptoError> {
    let cipher = Aes256Gcm::new(key.into());
    let mut iv = [0u8; IV_LEN];
    getrandom::getrandom(&mut iv).map_err(|_| CryptoError::Rng)?;
    let nonce = Nonce::from_slice(&iv);
    // aes-gcm returns ciphertext with the 16-byte tag appended → matches the
    // [ct][tag] tail of the wire frame.
    let ct = cipher
        .encrypt(nonce, Payload { msg: plaintext, aad: &[] })
        .map_err(|_| CryptoError::Encrypt)?;
    let mut out = Vec::with_capacity(IV_LEN + ct.len());
    out.extend_from_slice(&iv);
    out.extend_from_slice(&ct);
    Ok(out)
}

/// Decrypt an `[IV(12)] [ciphertext] [tag(16)]` frame.
pub fn decrypt(key: &[u8; 32], frame: &[u8]) -> Result<Vec<u8>, CryptoError> {
    if frame.len() < IV_LEN + TAG_LEN {
        return Err(CryptoError::FrameTooShort);
    }
    let cipher = Aes256Gcm::new(key.into());
    let (iv, ct_and_tag) = frame.split_at(IV_LEN);
    let nonce = Nonce::from_slice(iv);
    cipher
        .decrypt(nonce, Payload { msg: ct_and_tag, aad: &[] })
        .map_err(|_| CryptoError::Decrypt)
}

#[derive(Debug, thiserror::Error)]
pub enum CryptoError {
    #[error("rng failure")]
    Rng,
    #[error("remote public key must be 32 bytes")]
    BadPublicKey,
    #[error("hkdf expand failed")]
    Hkdf,
    #[error("aes-gcm encrypt failed")]
    Encrypt,
    #[error("aes-gcm decrypt/auth failed")]
    Decrypt,
    #[error("frame shorter than 28 bytes")]
    FrameTooShort,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ecdh_both_sides_agree() {
        let a = KeyPair::generate().unwrap();
        let b = KeyPair::generate().unwrap();
        let ka = a.derive_session_key(&b.public_raw).unwrap();
        let kb = b.derive_session_key(&a.public_raw).unwrap();
        assert_eq!(ka, kb, "both ends must derive the same session key");
    }

    #[test]
    fn encrypt_decrypt_roundtrip() {
        let a = KeyPair::generate().unwrap();
        let b = KeyPair::generate().unwrap();
        let key = a.derive_session_key(&b.public_raw).unwrap();

        let msg = br#"{"jsonrpc":"2.0","id":1,"result":{"pong":true}}"#;
        let frame = encrypt(&key, msg).unwrap();
        // frame = IV(12) + ct + tag(16)
        assert!(frame.len() >= IV_LEN + TAG_LEN);
        assert_eq!(frame.len(), IV_LEN + msg.len() + TAG_LEN);

        let out = decrypt(&key, &frame).unwrap();
        assert_eq!(out, msg);
    }

    #[test]
    fn each_frame_has_unique_iv() {
        let a = KeyPair::generate().unwrap();
        let b = KeyPair::generate().unwrap();
        let key = a.derive_session_key(&b.public_raw).unwrap();
        let f1 = encrypt(&key, b"hello").unwrap();
        let f2 = encrypt(&key, b"hello").unwrap();
        assert_ne!(&f1[..IV_LEN], &f2[..IV_LEN], "IVs must be random per frame");
    }

    #[test]
    fn tampered_frame_fails_auth() {
        let a = KeyPair::generate().unwrap();
        let b = KeyPair::generate().unwrap();
        let key = a.derive_session_key(&b.public_raw).unwrap();
        let mut frame = encrypt(&key, b"secret").unwrap();
        let last = frame.len() - 1;
        frame[last] ^= 0xff; // flip a tag byte
        assert!(matches!(decrypt(&key, &frame), Err(CryptoError::Decrypt)));
    }

    #[test]
    fn short_frame_rejected() {
        let key = [0u8; 32];
        assert!(matches!(decrypt(&key, &[0u8; 10]), Err(CryptoError::FrameTooShort)));
    }

    #[test]
    fn rejects_wrong_length_public_key() {
        let a = KeyPair::generate().unwrap();
        assert!(matches!(a.derive_session_key(&[0u8; 16]), Err(CryptoError::BadPublicKey)));
    }
}
